ml_genn/model.py

```diff
@@ -290,108 +303,229 @@ def convert_tf_model(tf_model, converter=Simple(),
+                    tf_in_layers.append(in_layer)
+            tf_in_layers_all[tf_layer] = tf_in_layers
 
-                layer.connect([previous_layer])
-                layer.set_weights(tf_layer.get_weights())
+            # find outbound layers
+            tf_out_layers = [n.outbound_layer for n in tf_layer.outbound_nodes
+                             if n in tf_model_nodes]
+            tf_out_layers_all[tf_layer] = tf_out_layers
 
-                model.layers.append(layer)
-                previous_layer = layer
-                pool_layer = None
 
-            # === AveragePooling2D Layers ===
-            elif isinstance(tf_layer, tf.keras.layers.AveragePooling2D):
-                print('deferring AveragePooling2D layer <{}>'.format(tf_layer.name))
+        # Perform any pre-compilation tasks
+        pre_compile_output = converter.pre_compile(tf_model)
 
-                pool_layer = tf_layer
+        # configure ML GeNN model build process
+        mlg_model_inputs = []
+        mlg_model_outputs = []
+        mlg_layer_lookup = {}
+        new_tf_layers = set()
+        traversed_tf_layers = set()
+
+        if isinstance(tf_model, tf.keras.models.Sequential):
+            tf_in_layers = tf_in_layers_all[tf_model.layers[0]]
+            assert(len(tf_in_layers) == 1)
+            tf_out_layers = [n.outbound_layer for n in tf_in_layers[0].outbound_nodes
+                             if n in tf_model_nodes]
+            tf_out_layers_all[tf_in_layers[0]] = tf_out_layers
+
+        else:
+            tf_in_layers = [tf_model.get_layer(name) for name in tf_model.input_names]
+
+        # === Input Layers ===
+        for tf_layer in tf_in_layers:
+            new_tf_layers.add(tf_layer)
+
+            print('configuring Input layer <{}>'.format(tf_layer.name))
+
+            name = tf_layer.name
+            assert(len(tf_layer.input_shape) == 1)
+            shape = tf_layer.input_shape[0][1:]
+
+            # create layer
+            mlg_layer = InputLayer(name=name, shape=shape,
+                neurons=converter.create_input_neurons(pre_compile_output))
+
+            mlg_layer_lookup[tf_layer] = mlg_layer
+            mlg_model_inputs.append(mlg_layer)
+
+        # while there are still layers to traverse
+        while new_tf_layers:
+            new_tf_layer = new_tf_layers.pop()
+            traversed_tf_layers.add(new_tf_layer)
+
+            # get next TF layer to configure
+            for tf_layer in tf_out_layers_all[new_tf_layer]:
+                tf_in_layers = set(tf_in_layers_all[tf_layer])
+                tf_out_layers = set(tf_out_layers_all[tf_layer])
+
+                # skip if we still need to configure inbound layers
+                if not traversed_tf_layers.issuperset(tf_in_layers):
+                    continue
+
+                # add this layer to new layers list
+                new_tf_layers.add(tf_layer)
+
+                # traverse ignored layers to find more inputs
```

Comment:
I don't fully understand the logic of this function; it is searching through all input layers and for those of the "ignored type" it is adding their inputs?
But the layers of ignored type remain in the set as well? They are only not added if they are the inputs to already ignored layers? - It seems rather fine-grained and convoluted. Is it really necessary and the best way of doing it?

Reply:
This part traverses through incoming Add, Flatten, Dropout layers to find actual weighted layers to connect to. Only non-ignored (i.e. weighted and pooling) layers are added to `tf_in_layers`. It is necessary, since there maybe multiple ignored layers to traverse before we find any good layers.

Reply:
Ok, thanks ... that makes sense but does it matter that the original layers of ignored type stay in the list ```tf_in_layers```? Is that intentional?

Reply:
They are deleted in line 420: `tf_in_layers.discard ...`.

Reply:
I just checked the code, and no ignored layers are present in `tf_in_layers` after this point. If you have some counterexample code, I would be very interested!

Reply:
Ok ... I see now ... it's ok. I find the logic a bit convoluted tbh 
It would be slightly more straightforward if you started with an empty set
```final_tf_in_layers``` and then added into it whenever a layer is "not ignored" rather than adding and discarding ... a bit like this:
```
 final_tf_in_layers = set()
     while tf_in_layers:
         tf_in_layer = tf_in_layers.pop()
         if isinstance(tf_in_layer, ignored_tf_layers):
             tf_in_layers.update(tf_in_layers_all[tf_in_layer])
         else:
             final_tf_in_layers.add(tf_in_layer)
tf_in_layers= final_tf_in_layers```

Reply:
Done.